### PR TITLE
Hdf2ckl severity

### PR DESCRIPTION
**Pull request by kemley76**
Fix to #5842.

### CODE REVIEW

**Comment by ejaronne** [comment]
The hdf2ckl export works properly, however, the Details tab should be:

Severity:
(tag.severity value)

Impact:
(Impact value)

**Review by georgedias — DISMISSED** [review]
Please see my inline comments

**Comment by kemley76** [comment]
I have made more changes regarding this issue. This includes:
- adding a severity tag upon importing a Checklist file
- using a control's severity tag to compute its severity rather than defaulting to use only the impact value
- compute the amount of "impact bubbles" in the results view with the impact rather than the severity
- be able to sort the controls by impact when clicking the impact column header
- be able to show falsy values in the details (like when there is impact of 0)

Done in previous commit, but not listed out
- use severity tag (if available) when exporting form HDF to CKL

This shows the sorting and bubble usage. The bubbles indicate impact, the text indicates severity. 
Screenshot 2024-06-05 at 11 48 52 AM

File: libs/hdf-converters/src/ckl-mapper/checklist-jsonix-converter.ts [hunk 1]
```diff
@@ -576,7 +576,14 @@ export class ChecklistJsonixConverter extends JsonixIntermediateConverter<
     }
   }
 
-  severityMap(impact: number): Severity {
+  severityMap(impact: number, severityTag: string): Severity {
+    severityTag = severityTag.toLowerCase();
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/src/ckl-mapper/checklist-jsonix-converter.ts:580]: I looked into various ways of doing this, but I landed on this as the most reasonable looking. If I don't include the `as string[]`, TypeScript complains that I am searching for a `string` in a `Severity` array.

**Review by Amndeep7 — CHANGES_REQUESTED** [review]
Currently the behavior is severity if it's there, otherwise fallback to impact.

What happens if there is a security override coming from a ckl?  What should the behavior be then?  I think the behavior should be the override if it's there, then the severity, then fallback to impact.

What should we do about the severity override justification?  Should that be shown in the details tab?

Questions for @ejaronne

**Comment by kemley76** [comment]
The original issue was caused by two things:
1. When a control has a severity tag, Heimdall didn't care and computed the severity it displays with the impact, which can differ from this tag
2. When exporting from hdf to checklist, the severity in the resulting checklist was computed with the hdf's impact and disregarded any severity tags. 
These have been addressed

Additional fixes that were addressed that may be beyond the scope of the original issue, but still needed to be addressed:

Compute a control's severity based on 
1. Any severity override (checklist vuln tag or hdf control tag)
2. Any severity tag (checklist tag or hdf control tag)
4. Impact 

Display severity override information in results table
Screenshot 2024-06-14 at 4 31 46 PM
Display severity override information in details view
Screenshot 2024-06-14 at 4 32 37 PM

It was also noticed that these changes affected the center "Severity Counts" ring graph. Before the "severities" in the graph were computed directly from the impact. Now, if there is a severity tag or severity override, that will be shown in the graph instead. When importing checklists, it is now impossible to see severity: none in this graph since severity: none does not exist in the checklist world.

**Comment by kemley76** [comment]
One last behavioral change was made here:
Upon exporting form hdf to checklist, severity tags none and critical now map to low and high respectively. 
This has the side effect that exporting from hdf to ckl and then back to hdf could result in different severity levels if they were originally none or critical.

**Comment by kemley76** [comment]
The impact and severity columns have now been separated and are being displayed with color coded chips.
Results View Changes

**Comment by ejaronne** [comment]
@kemley76 will you be coding a "filter to results with **severity overrides**", similar to the filer to profile errors?

**Comment by kemley76** [comment]
@ejaronne I previously looked into this and I think it might be too much for this issue. Currently, the system with the "filter to (profile) errors" or "filter to waived (controls)" is set up under the exact same component system as with the four main colored filter bars (passed, failed, N/A, not reviewed). Since "overridden" is not really a control status, this would make it difficult to implement under that system. Another option might be to add this filtering capabilities to the search bar at the top or to the switches on the "results view data" table, but still, might not be the most straightforward thing to implement.

**Comment by ejaronne** [comment]
> @ejaronne I previously looked into this and I think it might be too much for this issue. Currently, the system with the "filter to (profile) errors" or "filter to waived (controls)" is set up under the exact same component system as with the four main colored filter bars (passed, failed, N/A, not reviewed). Since "overridden" is not really a control status, this would make it difficult to implement under that system. Another option might be to add this filtering capabilities to the search bar at the top or to the switches on the "results view data" table, but still, might not be the most straightforward thing to implement.

Sorry I meant cosmetically the same, but logic based on severity override situation.

**Review by Amndeep7 — CHANGES_REQUESTED** [review]
> comment about the behavior change re potential data loss when doing the mapping of none/low -> low and high/critical -> high

hdf->ckl should encode the original information in the findings/comments are like we do for other attributes so that the ckl->hdf can try reading that first and then fallback to the actual checklist value so that we can do the full circle properly

File: apps/frontend/src/components/cards/controltable/ControlRowHeader.vue [hunk 2]
```diff
@@ -39,25 +39,46 @@
...
-        <div v-else>
-          <CircleRating
-            :filled-count="severity_arrow_count(control.data.tags.severity)"
-            :total-count="4"
-          />
-          <br />
-          <v-divider class="mx-1" />
-          {{ (control.data.tags.severity || 'none').toUpperCase() }}
-        </div>
+        <v-tooltip v-if="'severityoverride' in control.data.tags" bottom>
```
Comment (Amndeep7) [inline apps/frontend/src/components/cards/controltable/ControlRowHeader.vue:52]: Might be worthwhile updating the styling for v-tooltip so that the background is opaque.  The translucent background sometimes makes it hard to read text in the tooltip that is overlaid on top of other text on the page.  An example that's happening right now is the tooltip on the severity override overlapping the title text.

This will probably need to be fixed application wide.

File: apps/frontend/tests/unit/Results.spec.ts
Comment (Amndeep7) [inline apps/frontend/tests/unit/Results.spec.ts:1]: Can you add a cypress test as well.  Add your testing ckl file to the embedded samples that heimdall has, and then validate that the none/low <-> low mapping stuff works out well, that the delta symbol and tooltips are all showing up properly, etc.

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

**Review by Amndeep7 — CHANGES_REQUESTED** [review]
Address Eugene's request for the popup bar to filter to just the overriden controls.  It would probably behave similar to the nist tag filter option where we are looking through the tags object.  In this case, probably just looking to see if severityoverride is a tag that exists and has a value in it.  The filter would then end up like `override: true/false` in the searchbar string.  Don't forget to update the searchbar tooltip.

File: apps/frontend/src/components/cards/controltable/ControlRowHeader.vue (hunk 2 above)
Comment (kemley76) [reply apps/frontend/src/components/cards/controltable/ControlRowHeader.vue]: Done

**Comment by kemley76** [comment]
I have created a way to filter to controls with severity overrides by filtering for the specific severity override tag. 
Screenshot 2024-06-24 at 3 48 03 PM

This involved adding support to filter controls by the presence of specific tags, which is now documented in the search bar help menu.

**Comment by kemley76** [comment]
### Everything that was changed

Updated a bunch of tests
- Filtering severity override test
- Ckl mapper and reverse mapper tests with severity overrides
- Cypress test to ensure that severity overrides are rendered in results table

Changed mapper behavior
- when loading an HDF file, inspecJS now uses severity tag to compute severity if present
- when loading a ckl file
	- severity and severity override are used to compute impact 
	- severity override tags are added if severity override is given
	- severity override justification tag is added if in checklist
	- severity tag is always added now
- when converting hdf to ckl
	- severity tag is used to compute severity (falls back to use impact if no tag present)
	- severity override tag is used to compute severity override in checklist
	- if it wouldn’t be able to go back to hdf properly, the data is stored in checklist's third-party tools section

Updated Results Table
- Changed impact column to severity column (now on the right of title)
	- shows essentially the same information (bubbles + word -> colored word in chip)
	- information is only different for user if input file has a severity tag that differs from impact (either from mapping that adds one or from inspec profile with explicit tag)
	- if there is a severity override (currently only comes from checklist files, it shows a delta symbol with a tooltip that shows what it changed from and the justification)
- details page
	- if there is a severity tag, it is shown (falls back to internal severity which is typically computed from impact)
		- this makes it clear when there is a severity tag and severity override tag and what the difference is
	- fixed bug where impact 0 would not show up in details view

Filtering
- added ability to filter by the presence of a tag (tags:”severityoverride,nist”)
- added popup (like profile errors warning) that appears when a loaded profile has requirements with severity overrides (there is button to filter to them)

**Review by Amndeep7 — CHANGES_REQUESTED** [review]
At the very least, critical severities are not coming back as critical when you take the acme sample -> ckl -> load into heimdall.

File: apps/frontend/tests/unit/Results.spec.ts
Comment (kemley76) [reply apps/frontend/tests/unit/Results.spec.ts:1]: this was done

**Comment by kemley76** [comment]
> At the very least, critical severities are not coming back as critical when you take the acme sample -> ckl -> load into heimdall.

Should be resolved now

File: apps/frontend/src/components/cards/controltable/ControlRowDetails.vue [hunk 3]
```diff
@@ -202,7 +202,23 @@ export default class ControlRowDetails extends mixins(HtmlSanitizeMixin) {
     detailsMap.set('Caveat', this.control.hdf.descriptions.caveat);
     detailsMap.set('Desc', this.control.data.desc);
     detailsMap.set('Rationale', this.control.hdf.descriptions.rationale);
-    detailsMap.set('Severity', this.control.root.hdf.severity);
+    // default to showing severity tag, otherwise show the computed severity (based of impact or severityoverride)
```
Comment (Amndeep7) [inline apps/frontend/src/components/cards/controltable/ControlRowDetails.vue:205]:
```suggestion
    // default to showing severity tag, otherwise show the computed severity (based on impact or severityoverride)
```

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: libs/inspecjs/src/compat_impl/compat_inspec_1_0.ts [hunk 4]
```diff
@@ -164,6 +164,17 @@ abstract class HDFControl10 implements HDFControl {
   private static compute_severity(
     raw: ResultControl_1_0 | ProfileControl_1_0
   ): Severity {
+    const severities = ['none', 'low', 'medium', 'high', 'critical'];
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/inspecjs/src/compat_impl/compat_inspec_1_0.ts:167]: Done

File: libs/hdf-converters/test/mappers/reverse/checklist_reverse_mapper.spec.ts [hunk 5]
```diff
@@ -117,3 +117,33 @@ describe('Small RHEL8 HDF file', () => {
...
+    // );
+
+    const expected = fs.readFileSync(
+      'sample_jsons/checklist_mapper/converted-rhel7_overrides.ckl',
+      'utf-8'
+    );
+    const converted = mapper.toCkl();
+
+    expect(converted).toEqual(
+      expected.replace(/2\.10\.8/gi, hdfConvertersVersion)
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/test/mappers/reverse/checklist_reverse_mapper.spec.ts:146]: Done

File: libs/hdf-converters/src/ckl-mapper/checklist-mapper.ts [hunk 6]
```diff
@@ -52,60 +53,102 @@ function nistTag(input: string): string[] {
...
+      'hdfSpecificData.severityoverride',
+      severityOverride
+    );
   }
-  return vuln.severity;
+  return severityOverride;
 }
 
-function isJsonString(str: string): boolean {
+function parseJson(str: string): Result<any, Error> {
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/src/ckl-mapper/checklist-mapper.ts:97]: Done.

[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/src/ckl-mapper/checklist-mapper.ts:97]: Throughout hdf-converters, the return type of `JSON.parse` is considered to be `Result<string, unknown>`, so I just used that.

File: libs/hdf-converters/src/ckl-mapper/checklist-mapper.ts [hunk 7]
```diff
@@ -52,60 +53,102 @@ function nistTag(input: string): string[] {
...
 
+/**
+ * Function to find the computed severity of the given vuln
+ * with order of prescedence as:
+ * thirdPartyTools.hdfSpecificData.severityoverride, severityoverride,
+ * thidPartyTools.hdfSpecificData.severity, severity
+ * @param vuln - checklist vulnerability object
+ * @returns severity - string none, low, medium, high, critical
+ */
+function computeSeverity(vuln: ChecklistVuln): string {
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/src/ckl-mapper/checklist-mapper.ts:136]: done

File: libs/hdf-converters/src/ckl-mapper/checklist-jsonix-converter.ts [hunk 8]
```diff
@@ -650,6 +698,31 @@ export class ChecklistJsonixConverter extends JsonixIntermediateConverter<
       : '';
   }
 
+  // computes what the impact would be based on the given tags
+  computeImpact(
```
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/src/ckl-mapper/checklist-jsonix-converter.ts:702]: It turns out that this functionality (computing impact from severity) is not a inspecJs thing. It is, however, done a lot within the mappers through the base converter's `impactMapping` function. In this case, it also doesn't matter what critical maps to (.9 was chosen because it seems to be most common), because this value is only ever used to convert from ckl `thirdPartyTools.hdfSpecificData.severity` to `control.impact`, but if that mapping is wrong, there would be a `thirdPartyTools.hdfSpecificData.impact` to correct for it. The appropriate changes were made.

File: apps/frontend/src/components/cards/controltable/ControlRowHeader.vue [hunk 9]
```diff
@@ -41,23 +41,36 @@
...
-        <div v-if="showImpact">
-          <CircleRating
-            :filled-count="severity_arrow_count(control.hdf.severity)"
-            :total-count="4"
-          />
-          <v-divider class="mx-1" />
+        <v-tooltip v-if="'severityoverride' in control.data.tags" bottom>
+          <template #activator="{on}">
+            <span v-on="on">
+              <v-chip active-class="NONE" outlined :color="severity_color">
```
[the remark this answers is not in the document]
Comment (kemley76) [reply apps/frontend/src/components/cards/controltable/ControlRowHeader.vue:47]: Absolutely nothing. has been removed

File: apps/frontend/src/components/cards/controltable/ControlRowHeader.vue [hunk 10]
```diff
@@ -41,23 +41,36 @@
...
+                <v-icon size="16" class="mr-1" data-cy="severityOverride"
+                  >mdi-delta</v-icon
```
[the remark this answers is not in the document]
Comment (kemley76) [reply apps/frontend/src/components/cards/controltable/ControlRowHeader.vue:49]: done

File: apps/frontend/src/components/cards/controltable/ControlRowDetails.vue (hunk 3 above)
Comment (kemley76) [reply apps/frontend/src/components/cards/controltable/ControlRowDetails.vue:205]: done

File: libs/hdf-converters/sample_jsons/checklist_mapper/converted-rhel8_sample_oneOfEachControlStatus.ckl
[the remark this answers is not in the document]
Comment (kemley76) [reply libs/hdf-converters/sample_jsons/checklist_mapper/converted-rhel8_sample_oneOfEachControlStatus.ckl:1]: Done. https://github.com/mitre/saf/pull/2716

**Review by Amndeep7 — APPROVED** [review]

**Review by georgedias — APPROVED** [review]
ready to merge
